src/Storages/MergeTree/MergeTreeIndexGranuleBloomFilter.cpp

```diff
@@ -40,42 +24,28 @@ MergeTreeIndexGranuleBloomFilter::MergeTreeIndexGranuleBloomFilter(size_t bits_p
 }
 
 MergeTreeIndexGranuleBloomFilter::MergeTreeIndexGranuleBloomFilter(
-    size_t bits_per_row_, size_t hash_functions_, size_t total_rows_, const Blocks & granule_index_blocks_)
-        : total_rows(total_rows_), bits_per_row(bits_per_row_), hash_functions(hash_functions_)
+    size_t bits_per_row_, size_t hash_functions_, const std::vector<HashSet<UInt64>>& column_hashes_)
+        : bits_per_row(bits_per_row_), hash_functions(hash_functions_), bloom_filters(column_hashes_.size())
 {
-    if (granule_index_blocks_.empty() || !total_rows)
+    if (column_hashes_.empty())
         throw Exception("LOGICAL ERROR: granule_index_blocks empty or total_rows is zero.", ErrorCodes::LOGICAL_ERROR);
 
-    assertGranuleBlocksStructure(granule_index_blocks_);
-
-    for (size_t index = 0; index < granule_index_blocks_.size(); ++index)
-    {
-        Block granule_index_block = granule_index_blocks_[index];
+    size_t bloom_filter_max_size = 0;
+    for (const auto & column_hash : column_hashes_)
+        bloom_filter_max_size = std::max(bloom_filter_max_size, column_hash.size());
 
-        if (unlikely(!granule_index_block || !granule_index_block.rows()))
-            throw Exception("LOGICAL ERROR: granule_index_block is empty.", ErrorCodes::LOGICAL_ERROR);
-
-        if (index == 0)
-        {
-            static size_t atom_size = 8;
-
-            for (size_t column = 0, columns = granule_index_block.columns(); column < columns; ++column)
-            {
-                size_t total_items = total_rows;
-
-                if (const auto * array_col = typeid_cast<const ColumnArray *>(granule_index_block.getByPosition(column).column.get()))
```

Comment:
I'm not sure this condition would ever be true considering `granule_index_block` is initialised as [ColumnUInt64::create()](src/Interpreters/BloomFilterHash.h:130)